docker/nautilus.sh

```diff
@@ -0,0 +1,257 @@
+                                   NOTE: docker volumes will be created here to create persistence
+    -t, --tezos                    stops and removes existing tezos container if it exists
+                                   and rebuilds and starts the tezos container
+    -v, --volume                   REMOVES postgres volume and postgres volume folder, use at own risk as this will
+                                   require all blocks to be replaced in the postgres database, this is necessary if
+                                   there has been a schema change as simply rebuilding the container will not replace the schema
+
+Examples:
+    $CMD -a -p /$HOME/production-environment-1
+                                   # build, initialize, and run docker containers
+                                   # for conseil, postgres, and tezos
+                                   # takes config files from production-environment-1 folder
+
+
+Report bugs to <swap@cryptonomic.tech>.
+EOF
+}
+
+# parse command line arguments
+SHORT_OPTS='ab:cdh:p:tv'
+LONG_OPTS='all,build-name:,conseil,database,help,path-to-config:,tezos,volume'
+ARGS=$(getopt -o $SHORT_OPTS -l $LONG_OPTS -n "$CMD" -- "$@" 2>/dev/null)
+#check getopt command failure
+(( $? != 0 )) && fatal "invalid options"
+eval set -- "${ARGS}"
+# set execution flags and/or execute functions for each option
+while true ; do
+    case "$1" in
+        -a|--all) CONSEIL=1 ; POSTGRES=1 ; TEZOS=1 ; shift ;;
+        -b|--build-name) build_name="$2" ; shift 2 ;;
+	    -c|--conseil) CONSEIL=1 ; shift ;;
+        -d|--database) POSTGRES=1 ; shift ;;
+        -h|--help) display_usage && exit 0 ; shift ;;
+        -p|--path-to-config) path_to_config="$2" ; shift 2 ;;
+        -t|--tezos) TEZOS=1 ; shift ;;
+        -v|--volume) VOLUME=1 ; shift ;;
+        --) shift ; break ;;
+    esac
+done
+
+# ensure necessary command line parameters were specified(man test to see usages, checks contents of a string)
+[[ -z "${CONSEIL}${POSTGRES}${TEZOS}${VOLUME}" ]] && display_usage \
+    && fatal "Please specify at least one container type (examples: -a,-c,-d,-t)."
+
+
+build_time=$(date "+%Y.%m.%d")
+#sets current directory to variable DIR
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+#config file path, should contain exact files and directory structure as that of config/local
+PATH_TO_CONFIG="${path_to_config:-$DIR/config/local}"
+
+#working directory
+BUILD_NAME="${build_name:-nautilus_build_"$build_time"}"
+WORKING_DIR=$HOME/"$BUILD_NAME"
+#example: production-environment-1, will default to base folder name of config files
+DEPLOYMENT_ENV="$(basename "$PATH_TO_CONFIG")"
+#make working directory
+[[ -d "$WORKING_DIR" ]] || mkdir "$WORKING_DIR"
+#create variable for tezos-network(alphanet or mainnet)
+tezosnetwork=`cat "$PATH_TO_CONFIG"/tezos/tezos_network.txt`
+
+docker network create nautilus
+
+build_conseil () { 
```

Comment:
Each of these functions should accept parameters and only use these parameters in their logic. There should be no references at all to external variables.

Reply:
bash is not a higher level language and I don't think it has that capability.  After researching, I have not found any examples that show this ability.

Reply:
It does have it but it's awkward - https://stackoverflow.com/questions/6212219/passing-parameters-to-a-bash-function. 

You can get around the awkwardness by documenting for each function the parameters it takes using comments.

Reply:
what's the benefit for doing it this way?

Reply:
More modular code with fewer bugs. As close to functional programming as possible. We preclude any possibility of mixing contexts by separating local parameters from global variables.

This would entail something like:

```bash
bake_cake () {
# Whether to use sugar. If false, will use stevia.
PARAM_USE_SUGAR = $1 
# Temperature to set oven to
PARAM_OVER_TEMPERATURE = $2
# Bunch of code ensues
}

Reply:
done